sdlib/d/gc/arena.d

```diff
@@ -374,6 +423,32 @@ private:
 		unusedExtents.insert(e);
 	}
 
+	void shrinkAllocImpl(Extent* e, uint n, uint delta) {
+		assert(mutex.isHeld(), "Mutex not held!");
+
+		uint oldPages = cast(uint) e.size / PageSize;
+
+		assert(n <= PageCount - oldPages, "Invalid index!");
+		assert(delta > 0 && delta < oldPages, "Invalid delta!");
+
+		auto hpd = e.hpd;
+		auto prevLfr = hpd.longestFreeRange;
+
+		auto newPages = oldPages - delta;
+		e.at(e.address, newPages * PageSize, hpd);
+		hpd.clear(n + newPages, delta);
+
+		if (prevLfr != hpd.longestFreeRange) {
+			if (prevLfr != 0) {
+				auto index = getFreeSpaceClass(prevLfr);
+				heaps[index].remove(hpd);
+				filter &= ~(ulong(heaps[index].empty) << index);
+			}
+
+			registerHPD(hpd);
+		}
```

Comment:
I don't think this is correct. First several lfr can lead to the same bucket, but second, if the hpd was full, then it wouldn't be in any bucket.

The logic to handle this is already present when releasing pages, and it would be wise to see if there isn't a way to factor things out somewhat, or at least to use the same logic.

Reply:
It'd be good to have some tests for where the hpd ends up being after shrinking, involving the hdp being full before shrinking.

Reply:
I spent a fair amount of time looking at what `freePagesImpl` is doing and comparing to this and understand why it was different.

The former always remove the hdp from the heap it is in and reinsert it if appropriate, while this doesn't. As it turns out, `freePagesImpl` is correct here, because even though we might end up back in the same heap, it's not a given that the changes we made have not affected ordering. And even if we can convince ourselves that it does not with the current implementation, we just setup a trap for ourselves down the road. Who is going to think about changing this when updating the priority with which we select hdp?